### ISSUE TITLE
[Automation] Automate Go Formatting and similar tasks

### ISSUE THREAD

**Issue by Olshansk** [pokt-network/pocket#175]
## Objective

Create a foundational infrastructure so it is to automate certain tasks on pre-commits or PRs to avoid doing repetitive tasks.

## Origin Document

In https://github.com/pokt-network/pocket/pull/171, we did a code review to follow up on another PR after it is merged to master. See https://github.com/pokt-network/pocket/pull/166#issuecomment-1223262358 for details.

This approach raised the need for being able to easily automated tasks, ranging from code formatting, to tending to TODOs, running tests, etc...

## Goals

- [ ] Create a foundation so it is easy to add/automate new tasks that should be used and followed by others on the team
- [ ] Pick a technology that can cover most of our use cases (github workflows, git recommit hooks, etc...)

## Deliverable

- [ ] Document (1 page or less) on how we can add new automations
- [ ] At a minimum, automate the following:
    - [ ] Running `go fmt` so the imports are always ordered
    - [ ] Add an ability to run manual checks to make sure things like `require.True` are not used in the same line as `bytes.Equal` enforcing users to use `require.Equal`. Note that an override option will be necessary.
    - [ ] Remove unnecessary newlines

## Non-goals / Non-deliverables

- Making sure this solution is generalizable to anything
- Automating all the different things we need

## General issue deliverables

- [ ] Update the appropriate CHANGELOG
- [ ] Update any relevant READMEs (local and/or global)
- [ ] Update any relevant global documentation & references
- [ ] If applicable, update the source code tree explanation
- [ ] If applicable, add or update a state, sequence or flowchart diagram using [mermaid](https://mermaid-js.github.io/mermaid/)

---

**Creator**: @Olshansk 
**Co-Owners**: @okdas

**Comment by Olshansk** [comment]
Discussed offline with @adshmh and we have a couple options here:
1. A github check that runs `gofmt` errors out on non-empty diffs
2. Add prerecommit hoooks. See https://github.com/dnephin/pre-commit-golang for an example

**Comment by PoktBlade** [comment]
Sounds like we need a linter?

**Comment by PoktBlade** [comment]
https://github.com/golangci/golangci-lint

I was noticing that we are enforcing some rules, i.e: https://github.com/pokt-network/pocket/issues/156 then we can introduce a linter to enforce these rules alongside ensuring that formatting is consistent

**Comment by Olshansk** [comment]
We have https://github.com/pokt-network/pocket/blob/main/Makefile#L89, but we want to automate it.

1. I'm sure there are better linters
2. I'd like to have the foundation to add our specific "rules" to the linter

**Comment by Olshansk** [comment]
@PoktBlade Added the `community` label in case you want to take it on.

**Comment by adshmh** [comment]
> Discussed offline with @adshmh and we have a couple options here:
> 
>     1. A github check that runs `gofmt` errors out on non-empty diffs
> 
>     2. Add prerecommit hoooks. See https://github.com/dnephin/pre-commit-golang for an example

Thanks. We can even do both: 
 - the pre-commit hooks will fix the formatting before the PR is opened.
 - the gofmt and git diff check will be one of the CI checks to catch any incorrect formats before PR is merged.

**Comment by Olshansk** [comment]
Transferring offline discussion with @okdas here.

A few tools we can use are:
* https://staticcheck.io
* https://github.com/quasilyte/go-ruleguard
* https://semgrep.dev/

As a first starting step, we can potentially just automate a few of the targets we have in our [Makefile](https://github.com/pokt-network/pocket/blob/main/Makefile). Specifically:
* `make gofmt`
* `make go_clean_deps`
* `make go_staticcheck`

Lmk what you think

**Comment by Olshansk** [comment]
@okdas Lmk if you think this is out of scope, but is there a way to block merging the PR if it contains `DISCUSS_IN_THIS_COMMIT` or `TODO_IN_THIS_COMMIT`?

**Comment by Olshansk** [comment]
@okdas Can you attach https://github.com/pokt-network/pocket/pull/190 to this issue?

**Comment by okdas** [comment]
> but is there a way to block merging the PR if it contains

There is certainly a way to do this! Do you want me to add it to #190 scope?

**Comment by Olshansk** [comment]
Let's do it in a separate issue after #190. Can you create a separate issue for it?

**Comment by Olshansk** [comment]
@okdas Here's an example of a small lint that I'm not sure "an off the shelf" solution exists for: https://github.com/pokt-network/pocket/pull/218

Do you mind seeing what the best approach for this would be?

**Comment by okdas** [comment]
I think these are actually covered by default linter https://staticcheck.io/docs/checks#S1028 (and S1025 above)

Though, `log.` might not be captured (and we are going to move to other logging lib). In any case, we can write a linter rule for it using https://go-ruleguard.github.io which is implemented in #190.
